File: btc.py
import requests
from datetime import datetime, timezone, timedelta

def get_bitcoin_address_details(address):
    url = f'https://api.blockcypher.com/v1/btc/main/addrs/{address}?includeSubaccounts=true&includeUnconfirmed=true'
    response = requests.get(url)
    return response.json() if response.status_code == 200 else {'error': 'Unable to fetch data'}

def get_btc_to_idr_exchange_rate():
    url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=idr"
    response = requests.get(url)
    data = response.json()
    return data.get('bitcoin', {}).get('idr', None) if response.status_code == 200 else None

def get_btc_to_usd_price():
    url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd"
    response = requests.get(url)
    data = response.json()
    return data.get('bitcoin', {}).get('usd', None) if response.status_code == 200 else None

def get_fear_and_greed_index():
    url = "https://api.alternative.me/fng/"
    response = requests.get(url)
    data = response.json()
    return int(data['data'][0].get('value', '0')) if response.status_code == 200 else None

def determine_status(index_value):
    if index_value < 25:
        return "Extreme Fear"
    elif 25 <= index_value < 50:
        return "Fear"
    elif 50 <= index_value < 75:
        return "Greed"
    else:
        return "Extreme Greed"

def get_mempool_fee_rate():
    url = "https://mempool.space/api/v1/fees/recommended"
    response = requests.get(url)
    data = response.json()
    return {
        'slow': data.get('hourFee', 'N/A'),
        'medium': data.get('halfHourFee', 'N/A'),
        'fast': data.get('fastestFee', 'N/A')
    } if response.status_code == 200 else None

def format_rupiah(amount_satoshi, btc_to_idr_rate):
    amount_idr = (amount_satoshi / 100_000_000) * btc_to_idr_rate
    return f"Rp. {amount_idr:,.0f}".replace(",", ".")

def main():
    try:
        with open("wallet.txt", "r") as file:
            addresses = file.read().splitlines()
    except FileNotFoundError:
        print("File 'wallet' tidak ditemukan.")
        return

    total_balance_satoshi = sum(
        get_bitcoin_address_details(address).get('final_balance', 0) for address in addresses
    )

    btc_to_idr_rate = get_btc_to_idr_exchange_rate()
    btc_to_usd_price = get_btc_to_usd_price()
    fee_rate = get_mempool_fee_rate()
    fear_and_greed_index = get_fear_and_greed_index()

    if btc_to_idr_rate is None or btc_to_usd_price is None:
        print("Tidak dapat memperoleh nilai tukar atau harga BTC.")
        return

    formatted_balance = f"{total_balance_satoshi:,}".replace(",", ".")
    balance_rupiah = format_rupiah(total_balance_satoshi, btc_to_idr_rate)

    print("\n----------------------------------")
    print("~Balance~")
    print(f"{formatted_balance} sat")
    print(f"{balance_rupiah}")

    if fear_and_greed_index is not None:
        status = determine_status(fear_and_greed_index)
        print("----------------------------------")
        print("~Fear & Greed Index~")
        print(f"Index: {fear_and_greed_index}")
        print(f"Status: {status}")

    if fee_rate:
        print("----------------------------------")
        print("~Mempool fee rate~")
        print(f"Slow: {fee_rate['slow']} sat/vB")
        print(f"Medium: {fee_rate['medium']} sat/vB")
        print(f"Fast: {fee_rate['fast']} sat/vB")

    print("----------------------------------")
    print("~Price~")
    print(f"BTC: $ {btc_to_usd_price:,.0f}")
    print("----------------------------------")
if __name__ == "__main__":
    main()

File: btc.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <iomanip>
#include <map>
#include <curl/curl.h>
#include "json.hpp"

using namespace std;
using json = nlohmann::json;

size_t WriteCallback(void* contents, size_t size, size_t nmemb, string* output) {
    size_t totalSize = size * nmemb;
    output->append((char*)contents, totalSize);
    return totalSize;
}

json fetchJsonData(const string& url) {
    CURL* curl;
    CURLcode res;
    string readBuffer;

    curl = curl_easy_init();
    if (!curl) {
        cerr << "cURL gagal diinisialisasi!" << endl;
        return nullptr;
    }

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);

    res = curl_easy_perform(curl);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) {
        cerr << "cURL Error: " << curl_easy_strerror(res) << endl;
        return nullptr;
    }

    try {
        return json::parse(readBuffer);
    } catch (json::parse_error& e) {
        cerr << "JSON Parsing Error: " << e.what() << endl;
        return nullptr;
    }
}

long getBitcoinBalance(const string& address) {
    string url = "https://api.blockcypher.com/v1/btc/main/addrs/" + address + "?includeSubaccounts=true&includeUnconfirmed=true";
    json jsonData = fetchJsonData(url);

    if (jsonData.is_null() || !jsonData.contains("final_balance")) {
        cerr << "Gagal mengambil saldo BTC untuk alamat: " << address << endl;
        return 0;
    }

    try {
        return jsonData["final_balance"].get<long>();
    } catch (...) {
        return stol(jsonData["final_balance"].get<string>());
    }
}

double getBTCtoIDR() {
    string url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=idr";
    json jsonData = fetchJsonData(url);

    if (jsonData.is_null() || !jsonData.contains("bitcoin") || !jsonData["bitcoin"].contains("idr")) {
        cerr << "Gagal mengambil harga BTC ke IDR" << endl;
        return 0;
    }

    try {
        return jsonData["bitcoin"]["idr"].get<double>();
    } catch (...) {
        return stod(jsonData["bitcoin"]["idr"].get<string>());
    }
}

double getBTCtoUSD() {
    string url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd";
    json jsonData = fetchJsonData(url);

    if (jsonData.is_null() || !jsonData.contains("bitcoin") || !jsonData["bitcoin"].contains("usd")) {
        cerr << "Gagal mengambil harga BTC ke USD" << endl;
        return 0;
    }

    try {
        return jsonData["bitcoin"]["usd"].get<double>();
    } catch (...) {
        return stod(jsonData["bitcoin"]["usd"].get<string>());
    }
}

int getFearAndGreedIndex() {
    string url = "https://api.alternative.me/fng/";
    json jsonData = fetchJsonData(url);

    if (jsonData.is_null() || jsonData["data"].empty()) {
        cerr << "Gagal mengambil Fear & Greed Index" << endl;
        return 0;
    }

    try {
        return jsonData["data"][0]["value"].get<int>();
    } catch (...) {
        return stoi(jsonData["data"][0]["value"].get<string>());
    }
}

map<string, int> getMempoolFeeRate() {
    string url = "https://mempool.space/api/v1/fees/recommended";
    json jsonData = fetchJsonData(url);

    if (jsonData.is_null()) {
        cerr << "Gagal mengambil data fee transaksi." << endl;
        return {{"slow", 0}, {"medium", 0}, {"fast", 0}};
    }

    try {
        return {
            {"slow", jsonData["hourFee"].get<int>()},
            {"medium", jsonData["halfHourFee"].get<int>()},
            {"fast", jsonData["fastestFee"].get<int>()}
        };
    } catch (...) {
        return {
            {"slow", stoi(jsonData["hourFee"].get<string>())},
            {"medium", stoi(jsonData["halfHourFee"].get<string>())},
            {"fast", stoi(jsonData["fastestFee"].get<string>())}
        };
    }
}

string formatNumber(long long number) {
    string numStr = to_string(number);
    int insertPosition = numStr.length() - 3;
    while (insertPosition > 0) {
        numStr.insert(insertPosition, ",");
        insertPosition -= 3;
    }
    return numStr;
}

string formatRupiah(long amountSatoshi, double btcToIDR) {
    double amountIDR = (amountSatoshi / 100000000.0) * btcToIDR;
    return "Rp. " + formatNumber((long long)amountIDR);
}

string getFearStatus(int index) {
    if (index < 25) return "Extreme Fear";
    if (index < 50) return "Fear";
    if (index < 75) return "Greed";
    return "Extreme Greed";
}

int main() {
    ifstream file(getenv("HOME") + string("/wallet.txt"));
    if (!file) {
        cerr << "File 'wallet.txt' tidak ditemukan." << endl;
        return 1;
    }

    vector<string> addresses;
    string line;
    while (getline(file, line)) {
        addresses.push_back(line);
    }
    file.close();

    long totalBalanceSatoshi = 0;
    for (const auto& address : addresses) {
        totalBalanceSatoshi += getBitcoinBalance(address);
    }

    double btcToIDR = getBTCtoIDR();
    double btcToUSD = getBTCtoUSD();
    int fearAndGreedIndex = getFearAndGreedIndex();
    auto feeRates = getMempoolFeeRate();

    if (btcToIDR == 0 || btcToUSD == 0) {
        cerr << "Tidak dapat memperoleh nilai tukar atau harga BTC." << endl;
        return 1;
    }

    cout << "\n----------------------------------" << endl;
    cout << "~Balance~" << endl;
    cout << formatNumber(totalBalanceSatoshi) << " sat" << endl;
    cout << formatRupiah(totalBalanceSatoshi, btcToIDR) << endl;

    if (fearAndGreedIndex != 0) {
        cout << "----------------------------------" << endl;
        cout << "~Fear & Greed Index~" << endl;
        cout << "Index: " << fearAndGreedIndex << endl;
        cout << "Status: " << getFearStatus(fearAndGreedIndex) << endl;
    }

    if (!feeRates.empty()) {
        cout << "----------------------------------" << endl;
        cout << "~Mempool fee rate~" << endl;
        cout << "Slow: " << feeRates["slow"] << " sat/vB" << endl;
        cout << "Medium: " << feeRates["medium"] << " sat/vB" << endl;
        cout << "Fast: " << feeRates["fast"] << " sat/vB" << endl;
    }

    cout << "----------------------------------" << endl;
    cout << "~Price~" << endl;
    cout << "BTC: $ " << formatNumber((long long)btcToUSD) << endl;
    cout << "----------------------------------" << endl;

    return 0;
}
